pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+)
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block. It accepts the list of currently reserved
+// IPs and the requested CIDR as arguments and returns the /29 IP available in that CIDR
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// Parsing the CIDR resulted in an error
+func GetUnreservedIPBlock(reservedIPRanges map[string]bool, cidr string) (string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	buffer := bytes.NewBufferString("")
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, 8) {
+		overLap := false
+		buffer.WriteString(cidrIPBlock.String())
+		buffer.WriteString("/29")
+		for reservedIPRange := range reservedIPRanges {
+			err = validateCIDROverlap(buffer.String(), reservedIPRange)
+			if err != nil {
+				overLap = true
+				break
+			}
+		}
+		if !overLap {
+			return buffer.String(), nil
+		}
+		buffer.Reset()
+	}
+	return "", fmt.Errorf("All of the /29 IP ranges in the cidr %s are reserved", cidr)
+}
+
+// validateCIDROverlap checks if two cidrs have any overlapping IPs
+func validateCIDROverlap(cidr1 string, cidr2 string) error {
+	_, ipnet1, err := net.ParseCIDR(cidr1)
+	if err != nil {
+		return err
+	}
+
+	_, ipnet2, err := net.ParseCIDR(cidr2)
+	if err != nil {
+		return err
+	}
+
+	if ipnet1.Contains(ipnet2.IP) || ipnet2.Contains(ipnet1.IP) {
+		return fmt.Errorf("The cidr ranges %s and %s overlap", cidr1, cidr2)
+	}
+
+	return nil
+}
+
+// Increment the given IP value by the provided step
+func incrementIP(ip net.IP, step byte) error {
+	if uint8(255)-uint8(ip[len(ip)-1]) < uint8(step) {
+		return fmt.Errorf("IP overflow occured when incrementing ip %s with step %d", ip.String(), step)
+	}
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j] += step
```

Comment:
why are we incrementing each part of the IP by step then `breaking` if its over 0? This will always `break` after the first iteration unless the `step` is 0 and the original `ip` end was also 0.  So in practice this will only increment the last part of the ip because if step is non-zero the `break` will trigger.

Could you please clarify with comments or fix?

Reply:
It's to handle carry when rollover happens.